core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -727,11 +736,38 @@ private Expression translateParameter(RexDynamicParam expr,
     if (storageType == null) {
       storageType = typeFactory.getJavaClass(expr.getType());
     }
-    return nullAs.handle(
-        convert(
-            Expressions.call(root, BuiltInMethod.DATA_CONTEXT_GET.method,
-                Expressions.constant("?" + expr.getIndex())),
-            storageType));
+
+    Expression expression = Expressions.call(root, BuiltInMethod.DATA_CONTEXT_GET.method,
+            Expressions.constant("?" + expr.getIndex()));
+
+    Set<SqlTypeName> toIntegerTypes = Sets.newHashSet(
+            DATE,
+            TIME,
+            TIME_WITH_LOCAL_TIME_ZONE,
+            INTERVAL_YEAR,
+            INTERVAL_YEAR_MONTH,
+            INTERVAL_MONTH);
+
+    Expression midExpress = expression;
+    if (toIntegerTypes.contains(expr.getType().getSqlTypeName())) {
```

Comment:
intermediateExpression ? Also it's weird you call them toIntegerTypes but convert  to `String.class`.

Reply:
The outer call is `Expressions.call(BuiltInMethod.STRING_TO_DATE_ALLOW_NULLABLE.method ...` and the return type is integer. maybe i should change a more meaning name